runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/FusedPipelineTest.java

```diff
@@ -0,0 +1,143 @@
+    checkState(
+        fused.getFusedStages().size() == 2,
+        "Unexpected number of fused stages %s",
+        fused.getFusedStages());
+    RunnerApi.Pipeline fusedProto = fused.toPipeline(protoPipeline.getComponents());
+
+    assertThat(
+        "Root Transforms should all be present in the Pipeline Components",
+        fusedProto.getComponents().getTransformsMap().keySet(),
+        hasItems(fusedProto.getRootTransformIdsList().toArray(new String[0])));
+    assertThat(
+        "Should contain Impulse, GroupByKey, and two Environment Stages",
+        fusedProto.getRootTransformIdsCount(),
+        equalTo(4));
+    assertThat(fusedProto.getRootTransformIdsList(), hasItems("impulse", "gbk"));
+    assertRootsInTopologicalOrder(fusedProto);
+    // Since MapElements, WithKeys, and Values are all composites of a ParDo, we do prefix matching
+    // instead of looking at the inside of their expansions
+    assertThat(
+        "Fused transforms should be present in the components",
+        fusedProto.getComponents().getTransformsMap(),
+        allOf(hasKey(startsWith("map")), hasKey(startsWith("key")), hasKey(startsWith("values"))));
+    assertThat(
+        "Fused transforms shouldn't be present in the root IDs",
+        fusedProto.getRootTransformIdsList(),
+        not(hasItems(startsWith("map"), startsWith("key"), startsWith("values"))));
+
+    // The other components should be those of the original pipeline.
+    assertThat(
+        fusedProto.getComponents().getCodersMap(),
+        equalTo(protoPipeline.getComponents().getCodersMap()));
+    assertThat(
+        fusedProto.getComponents().getWindowingStrategiesMap(),
+        equalTo(protoPipeline.getComponents().getWindowingStrategiesMap()));
+    assertThat(
+        fusedProto.getComponents().getEnvironmentsMap(),
+        equalTo(protoPipeline.getComponents().getEnvironmentsMap()));
+    assertThat(
+        fusedProto.getComponents().getPcollectionsMap(),
+        equalTo(protoPipeline.getComponents().getPcollectionsMap()));
+  }
+
+  // For each transform in the root transforms, asserts that all consumed PCollections have been
+  // produced, and no produced PCollection has been consumed
+  private void assertRootsInTopologicalOrder(RunnerApi.Pipeline fusedProto) {
+    Set<String> consumedPCollections = new HashSet<>();
+    Set<String> producedPCollections = new HashSet<>();
+    for (int i = 0; i < fusedProto.getRootTransformIdsCount(); i++) {
+      PTransform rootTransform =
+          fusedProto.getComponents().getTransformsOrThrow(fusedProto.getRootTransformIds(i));
+      assertThat(
+          String.format(
+              "All %s consumed by %s must be produced before it",
+              PCollection.class.getSimpleName(), fusedProto.getRootTransformIds(i)),
+          producedPCollections,
+          hasItems(rootTransform.getInputsMap().values().toArray(new String[0])));
+      for (String consumed : consumedPCollections) {
```

Comment:
Don't all roots have zero consumed PCollections so this is just checking that no two roots produce the same PCollection.

Reply:
I see, not DAG roots.